charmcraft/commands/build.py

```diff
@@ -0,0 +1,281 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+
+import logging
+import os
+import pathlib
+import shutil
+import stat
+import subprocess
+import zipfile
+
+from charmcraft.cmdbase import BaseCommand, CommandError
+
+logger = logging.getLogger(__name__)
+
+# Some constants that are used through the code.
+CHARM_METADATA = 'metadata.yaml'
+BUILD_DIRNAME = 'build'
+VENV_DIRNAME = 'venv'
+
+# The file name and template for the dispatch script
+DISPATCH_FILENAME = 'dispatch'
+DISPATCH_CONTENT = """#!/bin/sh
+
+PYTHONPATH=lib:venv {entrypoint_relative_path}
+"""
+
+# The minimum set of hooks to be provided for compatibility with old Juju
+HOOK_NAMES = [
+    'install',
+    'start',
+    'upgrade-charm',
+]
+
+
+def polite_exec(cmd):
+    """Execute a command, only showing output if error."""
+    try:
+        proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+    except Exception as err:
+        logger.error("Execution crashed with %r for cmd %s", err, cmd)
+        return -1
+
+    for line in proc.stdout:
+        logger.debug(":: %s", line.rstrip())
+    retcode = proc.wait()
+
+    if retcode:
+        logger.error("Execution ended in %d for cmd %s", retcode, cmd)
+    return retcode
+
+
+def link(srcpath, destdir):
+    """Link a file."""
+    if not destdir.exists():
+        os.makedirs(destdir)
+
+    destpath = destdir / srcpath.name
+    destpath.symlink_to(srcpath)
+    return destpath
+
+
+class Builder:
+    """The package builder."""
+
+    def __init__(self, args):
+        self.charmdir = args['from']
+        self.entrypoint = args['entrypoint']
+        self.requirement_paths = args['requirement']
+
+        self.buildpath = self.charmdir / BUILD_DIRNAME
+
+    def run(self):
+        """Main building process."""
+        logger.debug("Building charm in %r", str(self.buildpath))
+
+        if self.buildpath.exists():
+            shutil.rmtree(str(self.buildpath))
+        self.buildpath.mkdir()
+
+        linked_entrypoint = self.handle_code()
+        self.handle_dispatcher(linked_entrypoint)
+        self.handle_dependencies()
+        zipname = self.handle_package()
+
+        logger.info("Done, charm left in %r", zipname)
+        return zipname
+
+    def handle_code(self):
+        """Handle basic files and the charm source code."""
+        # basic files
+        logger.debug("Linking in basic files and charm code")
+        link(self.charmdir / CHARM_METADATA, self.buildpath)
+
+        # the whole dir/tree if entry point is in a project's subdir, itself alone otherwise
+        if self.charmdir in self.entrypoint.parents and self.charmdir != self.entrypoint.parent:
+            # link the whole dir
+            linked_subdir = link(self.entrypoint.parent, self.buildpath)
```

Comment:
from a functional perspective, I find using link() everywhere a bit odd. As it means you cannot then use actual symlinks in the .charm because half of the things that are links are meant to be actual content, but some things would be meant as actual links. (eg build/hooks/install should likely be a symlink, but build/src/charm.py shouldn't)

Reply:
For now this will work, let's discuss in the future if we want some stuff to be copied and other symlinks, or what.